### ISSUE TITLE
assertion failed: imap/mboxevent.c: 743: filled_params(type, event)

### ISSUE THREAD

**Issue by akschu** [cyrusimap/cyrus-imapd#3115]
I'm trying to use external notifications on 3.2.2 but it doesn't work.

If I define

```
event_notifier: external
notify_external: /usr/cyrus/bin/cyrus_notify
event_groups: access
event_extra_params: clientAddress timestamp service

```
Then the imapd thread dies with this assertion:

Jul 15 18:01:54 snow imaps[28108]: Cannot notify event Login: missing
parameters: serverAddress clientAddress
Jul 15 18:01:54 snow imaps[28108]: Fatal error: Internal error:
assertion failed: imap/mboxevent.c: 743: filled_params(type, event)

If I remove event_groups and event_extra_params then notify never calls
my external script and notify breaks.

If I define "event_groups: access" and omit event_extra_params then I'm
back to:

Jul 15 18:14:26 snow imaps[28934]: Cannot notify event Login: missing
parameters: serverAddress
Jul 15 18:14:26 snow imaps[28934]: Fatal error: Internal error:
assertion failed: imap/mboxevent.c: 743: filled_params(type, event)

**Comment by elliefm** [comment]
Thanks.  I had a look at this briefly when you sent it to the mailing list, and there's some weirdness about "serverAddress" vs "serverDomain" that I don't yet understand.

None of our existing Cassandane tests test the "access" event group.  I tweaked my local Cassandane to enable the "access" event group in one of the existing event tests, but was unable to reproduce the assertion failure.  The Login event that was generated contained "serverDomain" and not "serverAddress"???  Not sure what's going on yet, but more investigation is definitely required!

I might end up needing more information about your configuration, but I don't know what to ask for yet.

**Comment by elliefm** [comment]
I don't suppose you're connecting over something other than a TCP/IP socket, are you?  I can see a case where this field is left blank if the connection isn't on an IP socket, like maybe a unix domain socket or something

**Comment by elliefm** [comment]
Sketching some things here: https://github.com/elliefm/cyrus-imapd/tree/v32/3115-login-event-serverclient-info

**Comment by akschu** [comment]
Thanks for working on this.

I assumed you wanted me to try this branch so I did, but same error:

Here is my event config:

# events
event_notifier: external
notify_external: /usr/cyrus/bin/perl_script_I_wrote_to_log_and_track_some_stuff
event_groups: access
event_extra_params: clientAddress timestamp service

My script is basically to just keep track of which users are actually using their mailbox by tracking their logins, it also sets up some permissions and other administrative tasks.

A few notes:

1.  This works fine on 3.0.13.  It's 3.2 that spits out the error.

2.  I'm using a little bit of a different ptloader config if that makes a difference:

lmtp_downcase_rcpt: yes
username_tolower: yes
sasl_pwcheck_method: saslauthd
sasl_mech_list: PLAIN LOGIN
allowapop: no
allowplaintext: yes
virtdomains: no
pts_module: ldap
auth_mech: pts
ldap_base: dc=domain,dc=net
ldap_sasl: no
ldap_bind_dn: cn=Manager,dc=domain,dc=net
ldap_password: topsecret
ldap_uri: ldap://localhost
ldap_filter: (|(uid=%u)(mailLocalAddress=%u))
ldap_user_attribute: uid
ptloader_sock: /var/spool/imap/socket/ptsock
ldap_size_limit: 100
ldap_member_method: attribute
ldap_member_attribute: member

I'm doing this because it take whatever username I get then resolves it against both the uid and email address to return the uid which we authenticate against.  This is because my imap system is old enough to be before the days of using an email address for a username, and because it allows for the user to authenticate as their username or any of their email addresses, which might be a few.

With those config options I get:

Jul 23 20:03:51 server imaps[25278]: accepted connection
Jul 23 20:03:51 server master[25300]: about to exec /usr/cyrus/libexec/imapd
Jul 23 20:03:51 server imaps[25278]: tls_client_ca_dir=(NULL) tls_client_ca_file=/etc/ssl/certs/imap.crt
Jul 23 20:03:51 server imaps[25278]: tls_server_cert=/etc/ssl/certs/imap.crt tls_server_key=/etc/ssl/certs/imap.key
Jul 23 20:03:51 server imaps[25278]: inittls: Loading hard-coded DH parameters
Jul 23 20:03:51 server imaps[25278]: Set client CA list: Client cert requested, not required
Jul 23 20:03:51 server imaps[25278]: TLS Server Name Indication (SNI) Extension: "server.net"
Jul 23 20:03:51 server imaps[25300]: executed
Jul 23 20:03:51 server imaps[25278]: SSL_accept() incomplete -> wait
Jul 23 20:03:51 server imaps[25300]: TLS is available.
Jul 23 20:03:51 server imaps[25278]: SSL_accept() succeeded -> done
Jul 23 20:03:51 server imaps[25278]: starttls: TLSv1.2 with cipher ECDHE-RSA-AES128-SHA (128/128 bits new) no authentication
Jul 23 20:03:51 server imaps[25278]: extractor_init(0x2122a90)
Jul 23 20:03:51 server imaps[25278]: TLS is available.
Jul 23 20:03:51 server imaps[25278]: command: 1 Authenticate
Jul 23 20:03:51 server imaps[25278]: ptload(): fetched cache record (user@domain.com)(mark 1595559593, current 1595559831, limit 1595549031)
Jul 23 20:03:51 server imaps[25278]: ptload returning data
Jul 23 20:03:51 server imaps[25278]: canonified user@domain.com -> user
Jul 23 20:03:51 server imaps[25278]: login: <ip address> user PLAIN+TLS User logged in SESSIONID=<server-1595559831-25278-1-3240825040740759235>
Jul 23 20:03:51 server imaps[25278]: TLS is available.
Jul 23 20:03:51 server imaps[25278]: Cannot notify event Login: missing parameters: serverAddress clientAddress
Jul 23 20:03:51 server imaps[25278]: Fatal error: Internal error: assertion failed: imap/mboxevent.c: 743: filled_params(type, event)
Jul 23 20:03:51 server imaps[25278]: extractor_destroy(0x216cc30)
Jul 23 20:03:51 server master[25274]: process type:SERVICE name:imaps path:/usr/cyrus/libexec/imapd age:7.731s pid:25278 exited, status 70
Jul 23 20:03:51 server master[25274]: service imaps/ipv4 pid 25278 in BUSY state: terminated abnormally

**Comment by akschu** [comment]
> I don't suppose you're connecting over something other than a TCP/IP socket, are you? I can see a case where this field is left blank if the connection isn't on an IP socket, like maybe a unix domain socket or something

Connecting to imap?  No, tcp socket with TLS.  I dug through the code a little and I couldn't figure out why serverAddress is missing.

**Comment by elliefm** [comment]
Ah, that's interesting.

> Jul 23 20:03:51 server imaps[25278]: TLS Server Name Indication (SNI) Extension: "server.net"

I wonder if there's something wrong with our support for this extension, since it seems to care about server names, and server names are the problem we're experiencing.

I thought I saw something that might have been related to TLS when I first looked at this, but when I looked closer I didn't think it was related after all, so I moved on.  But maybe I need to look there again...

I recently added the ability for our testing software (Cassandane) to use TLS in tests, but so far only one test uses it.  So one thing I'll try is adding a TLS-enabled test that also has the "access" event group enabled, and see what happens.  If I can manage to reproduce the assertion, then I can locally patch cyrus to abort() with a core dump on assertions (instead of exiting with a syslog message and error code), and get a backtrace from the core dump...

**Comment by akschu** [comment]
Ellie,

So sorry to lead you down that wild goose chase, but server.net is just me sanitizing the data before publicly posting.   The data on that line matches the dns resolvable hostname of the system that matches the CN of the cert.

Matt

**Comment by elliefm** [comment]
Yeah I figured that; the interesting detail to me is that it cares at all!

**Comment by elliefm** [comment]
(Sorry about double post; GitHub was telling me I couldn't post, but was clearly accepting them still, doh.)

**Comment by elliefm** [comment]
I've force-pushed my branch from earlier.  It doesn't have the sketches from earlier anymore (no longer needed), but does contain a single new commit which is hopefully the fix:  https://github.com/cyrusimap/cyrus-imapd/compare/cyrus-imapd-3.2...elliefm:v32/3115-login-event-serverclient-info

Can you please test this out and report back with results?  Thanks!

**Comment by akschu** [comment]
Yay!!!! that fixed it!  Thank you so much!

**Comment by elliefm** [comment]
That's great news, thanks for trying it out. :)  I'll get this merged and it'll be included in 3.2.3.

**Comment by elliefm** [comment]
Reopening this while I keep trying to get some tests up for it.  WIP (currently not working) is here: https://github.com/cyrusimap/cassandane/compare/master...elliefm:3115-mboxevent-access-tls

**Comment by elliefm** [comment]
Test is now working properly and merged upstream

**Comment by akschu** [comment]
Ellie, there are more issues.  This patch fixed the imap part, but not the pop3 part.  Also, the pop3 server issued another assert at logout because, "Cannot notify event Logout: missing parameters: clientAddress".

Attached is an expanded patch that includes your changes for the pop3 server as well as the imap server, along with the clientAddress part we need during logout on the pop3 side.

[3115-login-event-serverclient-info-UPDATED.txt](https://github.com/cyrusimap/cyrus-imapd/files/5024148/3115-login-event-serverclient-info-UPDATED.txt)

**Comment by elliefm** [comment]
Thanks, I didn't realise pop3 was also affected.  I'm merging your patch now, so it will also be included in 3.2.3 :)
